jobs/pywikibot/scripts/updatestatistics.py

```diff
@@ -143,6 +142,21 @@ def update_list_of_pages(template, pages):
         template_page.save('Updated on ' +
                            date.today().strftime('%B %d, %Y'))
 
+def deduplicate_lists(list_to_dedup, base_list):
+    '''
+    will return list_to_dedup after removing any element that exists in
+    base_list
+
+    will maintain the order of elements in list_to_dedup
+    '''
+
+    dict_from_list = OrderedDict.fromkeys(list_to_dedup)
+    for elem in base_list:
+        if elem in dict_from_list:
+            list_to_dedup.remove(elem)
```

Comment:
I assume you're doing the `OrderedDict` stuff to make the complexity O(m + n) instead of O(mn)? If so, doesn't this line make the complexity O(mn)? Should this be `dict_from_list.remove(elem)`?

More generally, I don't think it's worth doing these kinds of optimisations unless really necessary; they make the code harder to understand, more prone to bugs, and don't actually make it practically faster or better. A simple double loop would suffice IMO.

Reply:
Um, no, actually the complexity is still O(m+n) where m is the size of the `base_list` and n is the size of the list to be deduped `list_to_dedup`

1. Form a dictionary from value to bool (say) of the list to be deduplicated `list_to_dedup` - `O(n)`
1. Take every element in the base_list - `O(m)`
    1. Check if it is there in `list_to_dedup` using the dictionary - `O(1)`
    1. If it exists in that list, then remove it - `O(1)`

I did this because we can't iterate over `list_to_dedup` and remove elements from it at the same time.

This approach is faster than O(mn) and I would prefer to stick with it. I think this is a pretty standard approach? (forming a dictionary and removing duplicates) I will add some documentation and then you can see how you feel about the code.

Reply:
Isn't step 2. ii. `O(n)` since you're calling `remove` on a `list`?

Reply:
Ah, jeez, sorry sorry! I totally missed that. I looked at my code a few times to see why you thought it was O(mn), I completely missed that. Sorry about that! :cry: 

I have changed the code now. I start with an empty list and add elements that aren't in `base_list` to the new list and return the new list.